src/backtrace.rs

```diff
@@ -0,0 +1,11 @@
+use std::error::Error;
+
+pub trait Backtrace: Error {
+    fn backtrace(&self) -> Option<&backtrace::Backtrace>;
+}
```

Comment:
Currently, this is used in two places when `cfg(not(feature = "stdbacktrace"))`. It's used like so:

```
impl #impl_generics thiserror::Backtrace for #ty #ty_generics #where_clause {
    #backtrace_method
}
```

This doesn't _have_ to be done with a trait. It could instead be done by creating a `backtrace()` method directly for the type. Something like:

```
impl #impl_generics #ty #ty_generics #where_clause {
    pub #backtrace_method
}
```

In my testing, this wouldn't collide with an `impl` for the type that the user did themselves (as long as they don't also try to create a `backtrace()` method). E.g.:

```rust
#[derive(thiserror::Error, Debug)]
#[error("...")]
pub struct MyError {
    backtrace: Backtrace,
}

impl MyError {
    fn whatever() {
        // ...
    }
}   
```

The macros would expand to include _another_ `impl MyError` block for the `backtrace()` method. But Rust allows for [multiple `impl` blocks for the same type](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bc64d3909a826392844f8340ab543133), so that shouldn't be a problem.

The user also then wouldn't need to `use thiserror::Backtrace` to make the `backtrace()` method available from the trait.

Thoughts?

Reply:
Though, the trait would still be needed at least privately, to be able to pull the backtrace off a source error. That's currently used like this:

```
#varsource.as_ref().and_then(|source| source.as_dyn_error().downcast_ref::<&(dyn thiserror::Backtrace)>().and_then(|source| thiserror::Backtrace::backtrace(source)))
```